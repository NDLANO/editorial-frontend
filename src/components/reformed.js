/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * A modified version of https://github.com/davezuko/react-reformed (MIT license)
 */

import React from 'react';
import PropTypes from 'prop-types';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { getComponentName } from 'ndla-util';

const makeWrapper = WrappedComponent => {
  class FormWrapper extends React.Component {
    constructor(props, ctx) {
      super(props, ctx);
      this.state = {
        submitted: false,
        model: props.initialModel || {},
        fields: {},
      };
      this.setModel = this.setModel.bind(this);
      this.setProperty = this.setProperty.bind(this);
      this.bindToChangeEvent = this.bindToChangeEvent.bind(this);
      this.bindInput = this.bindInput.bind(this);
      this.setSubmitted = this.setSubmitted.bind(this);
    }

    componentWillReceiveProps(nextProps) {
      const currentModel = this.props.initialModel;
      const nextModel = nextProps.initialModel;

      if (this.props.resetOnInitialModelChange && currentModel !== nextModel) {
        const hasChanges = Object.keys(nextModel).find(
          key => nextModel[key] !== currentModel[key],
        );
        if (hasChanges) {
          this.setModel(nextModel);
        }
      }
    }

    setModel(model) {
      this.setState({ model });
      return model;
    }

    setSubmitted(submitted) {
      this.setState({ submitted });
    }

    setProperty(prop, value, isDirty = true) {
      const model = {
        ...this.state.model,
        [prop]: value,
      };

      if (this.state.fields[prop] && this.state.fields[prop].isDirty) {
        this.setState(prevstate => ({
          model: { ...prevstate.model, [prop]: value },
        }));
      } else {
        this.setState(prevstate => {
          const fields = { ...prevstate.fields, [prop]: { isDirty } };
          return { model: { ...prevstate.model, [prop]: value }, fields };
        });
      }
      return model;
    }

    bindToChangeEvent(e) {
      const { name, type, value } = e.target;
      if (type === 'checkbox') {
        const oldCheckboxValue = this.state.model[name] || [];
        const newCheckboxValue = e.target.checked
          ? oldCheckboxValue.concat(value)
          : oldCheckboxValue.filter(v => v !== value);

        this.setProperty(name, newCheckboxValue);
      } else if (type === 'EditorState') {
        this.setProperty(name, value, value.getCurrentContent().hasText()); // Only set dirty flag if text has changed
      } else if (type === 'SlateEditorState') {
        this.setProperty(name, value); // TODO: Handle dirty flag with SlateEditorState
      } else {
        this.setProperty(name, value);
      }
    }

    bindInput(name) {
      return {
        name,
        value: this.state.model[name] || '',
        onChange: this.bindToChangeEvent,
      };
    }

    render() {
      const nextProps = {
        ...this.props,
        bindInput: this.bindInput,
        bindToChangeEvent: this.bindToChangeEvent,
        model: this.state.model,
        fields: this.state.fields,
        submitted: this.state.submitted,
        setProperty: this.setProperty,
        setSubmitted: this.setSubmitted,
        setModel: this.setModel,
      };

      return React.createElement(WrappedComponent, nextProps);
    }
  }

  FormWrapper.propTypes = {
    initialModel: PropTypes.object, //eslint-disable-line
    resetOnInitialModelChange: PropTypes.bool.isRequired,
  };
  FormWrapper.defaultProps = {
    resetOnInitialModelChange: false,
  };
  FormWrapper.displayName = `Reformed(${getComponentName(WrappedComponent)})`;
  return hoistNonReactStatics(FormWrapper, WrappedComponent);
};

export default makeWrapper;
