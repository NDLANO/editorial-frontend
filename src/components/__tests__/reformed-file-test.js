/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import renderer from 'react-test-renderer';
import PropTypes from 'prop-types';
import reformed from '../reformed';

export const FileForm = props => {
  const { bindInput } = props;

  const fileInfo = bindInput('file').value;
  const filePath = bindInput('filepath').value;
  return (
    <form>
      <input type="file" {...bindInput('file', true)} />
      {fileInfo
        ? <span>
            {`${fileInfo.name} ${fileInfo.type} ${fileInfo.size}`}
          </span>
        : null}
      {filePath
        ? <span>
            {filePath}
          </span>
        : null}
    </form>
  );
};

FileForm.propTypes = {
  bindInput: PropTypes.func.isRequired,
};

test('reformed HOC renders file input correctly', () => {
  const Reformed = reformed(FileForm);

  const component = renderer.create(
    <Reformed initialModel={{ file: undefined }} />,
  );

  expect(component.toJSON()).toMatchSnapshot();
});

test('reformed HOC handles file input change correctly', () => {
  URL.createObjectURL = file => `blob:${file.name}`; // Mock
  const Reformed = reformed(FileForm);

  const component = renderer.create(
    <Reformed initialModel={{ file: undefined }} />,
  );
  const tree = component.toJSON();

  const input = tree.children[0];
  const file = new File(['foo'], 'foo.txt', { type: 'text/plain' });

  input.props.onChange({
    target: { files: [file], type: 'file', name: input.props.name },
  });

  expect(component.toJSON()).toMatchSnapshot();
});
