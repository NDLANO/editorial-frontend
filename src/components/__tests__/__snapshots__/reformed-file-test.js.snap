// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reformed HOC handles file input change correctly 1`] = `
<form>
  <input
    name="file"
    onChange={[Function]}
    type="file"
  />
  <span>
    foo.txt text/plain 3
  </span>
  <span>
    blob:foo.txt
  </span>
</form>
`;

exports[`reformed HOC renders file input correctly 1`] = `
<form>
  <input
    name="file"
    onChange={[Function]}
    type="file"
  />
</form>
`;
